# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: docshell/base/BrowsingContext.cpp
# Commit: 3b7697e19437
# Full Hash: 3b7697e19437fb8d4a63cefa69b87ed76bdd3ae3
# Author: Alexandra Borovova <aborovova@mozilla.com>
# Date: 2025-09-10 21:28:29
# Description:
#   Bug 1987777 - Reapply language and timezone overrides only when they exist and add the checks for WindowContext to avoid the crash. r=dom-core,smaug
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D264378
# ==============================================================================

diff -r 181eaac46885 -r 3b7697e19437 docshell/base/BrowsingContext.cpp
--- a/docshell/base/BrowsingContext.cpp	Wed Sep 10 11:52:56 2025 +0000
+++ b/docshell/base/BrowsingContext.cpp	Wed Sep 10 12:06:02 2025 +0000
@@ -3236,27 +3236,29 @@
   const nsCString& languageOverride = GetLanguageOverride();
 
   PreOrderWalk([&](BrowsingContext* aBrowsingContext) {
-    RefPtr<WindowContext> windowContext =
-        aBrowsingContext->GetCurrentWindowContext();
-
-    if (nsCOMPtr<nsPIDOMWindowInner> window = windowContext->GetInnerWindow()) {
-      JSObject* global = nsGlobalWindowInner::Cast(window)->GetGlobalJSObject();
-      JS::Realm* realm = JS::GetObjectRealmOrNull(global);
-
-      if (mDefaultLocale == nullptr) {
-        AutoJSAPI jsapi;
-        if (jsapi.Init(window)) {
-          JSContext* context = jsapi.cx();
-          mDefaultLocale = JS_GetDefaultLocale(context);
+    if (RefPtr<WindowContext> windowContext =
+            aBrowsingContext->GetCurrentWindowContext()) {
+      if (nsCOMPtr<nsPIDOMWindowInner> window =
+              windowContext->GetInnerWindow()) {
+        JSObject* global =
+            nsGlobalWindowInner::Cast(window)->GetGlobalJSObject();
+        JS::Realm* realm = JS::GetObjectRealmOrNull(global);
+
+        if (mDefaultLocale == nullptr) {
+          AutoJSAPI jsapi;
+          if (jsapi.Init(window)) {
+            JSContext* context = jsapi.cx();
+            mDefaultLocale = JS_GetDefaultLocale(context);
+          }
         }
-      }
-
-      if (languageOverride.IsEmpty()) {
-        JS::SetRealmLocaleOverride(realm, mDefaultLocale.get());
-        mDefaultLocale = nullptr;
-      } else {
-        JS::SetRealmLocaleOverride(realm,
-                                   PromiseFlatCString(languageOverride).get());
+
+        if (languageOverride.IsEmpty()) {
+          JS::SetRealmLocaleOverride(realm, mDefaultLocale.get());
+          mDefaultLocale = nullptr;
+        } else {
+          JS::SetRealmLocaleOverride(
+              realm, PromiseFlatCString(languageOverride).get());
+        }
       }
     }
   });
@@ -3619,18 +3621,20 @@
   MOZ_ASSERT(IsTop());
 
   PreOrderWalk([&](BrowsingContext* aBrowsingContext) {
-    RefPtr<WindowContext> windowContext =
-        aBrowsingContext->GetCurrentWindowContext();
-
-    if (nsCOMPtr<nsPIDOMWindowInner> window = windowContext->GetInnerWindow()) {
-      JSObject* global = nsGlobalWindowInner::Cast(window)->GetGlobalJSObject();
-      JS::Realm* realm = JS::GetObjectRealmOrNull(global);
-
-      if (GetTimezoneOverride().IsEmpty()) {
-        JS::SetRealmTimezoneOverride(realm, nullptr);
-      } else {
-        JS::SetRealmTimezoneOverride(
-            realm, NS_ConvertUTF16toUTF8(GetTimezoneOverride()).get());
+    if (RefPtr<WindowContext> windowContext =
+            aBrowsingContext->GetCurrentWindowContext()) {
+      if (nsCOMPtr<nsPIDOMWindowInner> window =
+              windowContext->GetInnerWindow()) {
+        JSObject* global =
+            nsGlobalWindowInner::Cast(window)->GetGlobalJSObject();
+        JS::Realm* realm = JS::GetObjectRealmOrNull(global);
+
+        if (GetTimezoneOverride().IsEmpty()) {
+          JS::SetRealmTimezoneOverride(realm, nullptr);
+        } else {
+          JS::SetRealmTimezoneOverride(
+              realm, NS_ConvertUTF16toUTF8(GetTimezoneOverride()).get());
+        }
       }
     }
   });